src/JSConfetti.ts

```diff
@@ -4,17 +4,59 @@ import { createCanvas } from './createCanvas'
 import { normalizeConfettiConfig } from './normalizeConfettiConfig'
 import { IPosition, IJSConfettiConfig, IAddConfettiConfig } from './types'
 
+class ConfettiBatch {
+  private promiseCompletion?: () => void
+  private promise?: Promise<void>
+  private members: ConfettiShape[]
+
+  constructor() {
+    this.members = []
+  }
+
+  getPromise(): Promise<void> {
+    if (!this.promise) {
+      this.promise = new Promise((completionCallback) => this.promiseCompletion = completionCallback)
+    }
+
+    return this.promise
+  }
+
+  addShapes(...shapes: ConfettiShape[]): void {
+    this.members.push(...shapes)
+  }
+
+  complete(): boolean {
+    if (this.members.length) {
+      return false
+    }
+    
+    // If a promise was never requested, we can't complete it
+    this.promiseCompletion?.()
+
+    return true
+  }
+
+  filterShapes(height: number): void {
+    if (this.members.length < 1) {
+      return
+    }
+
+    this.members = this.members.filter((shape) => shape.getIsVisibleOnCanvas(height))
```

Comment:
Is there a possibility not to filter confettis here and also in `loop()` method? It seems to me that these operations quite like each other and we can filter them just in one place

Maybe we can get rid of `JSConfetti.shapes`?

Reply:
Originally I was trying to stay away from a large refactor. But your comment pushed me to do so -- take a look now and let me know what you think. The commit message has some more details:

Reduce the number of passes through shapes/batches for rendering & cleaning up.

- Moved primary processing of *shapes* into the batch class itself
- Change loop to iterate over the batches, not shapes. Each batch handles the shape work
- Changed the work to `filter` not `forEach`: this minimizes the number of passes over the list, trading off against allocations of an array every time. Now there is `batch * shapes` operations every time with no `2 * shapes` every 100 iterations.